src/tpm2_akparse.c

```diff
@@ -138,38 +138,35 @@ int parseAKPublicArea()
 {
     TPM2B_PUBLIC outPublic;
     UINT16 size = sizeof(outPublic);
-    if( loadDataFromFile(akDataFile, (UINT8 *)&outPublic, &size) )
+    TPM2B *sizedBuffer;
+    int rc = -1;
+
+    if ( loadDataFromFile(akDataFile, (UINT8 *)&outPublic, &size) )
     {
         return -1;
     }
 
-    if( TPM_ALG_RSA == outPublic.t.publicArea.type )
-    {
-        if(SaveKeyToFile(akKeyFile, TPM_ALG_RSA, outPublic.t.publicArea.unique.rsa.t))
-            return -2;
-    }
-    else if( TPM_ALG_ECC == outPublic.t.publicArea.type )
-    {
-        if(SaveEccKeyToFile(akKeyFile, TPM_ALG_ECC, outPublic.t.publicArea.unique.ecc))
-            return -3;
-    }
-    else if( TPM_ALG_KEYEDHASH == outPublic.t.publicArea.type )
-    {
-        if(SaveKeyToFile(akKeyFile, TPM_ALG_KEYEDHASH, outPublic.t.publicArea.unique.keyedHash.t))
-            return -4;
-    }
-    else if( TPM_ALG_SYMCIPHER == outPublic.t.publicArea.type )
+    switch ( outPublic.t.publicArea.type )
     {
-        if(SaveKeyToFile(akKeyFile, TPM_ALG_SYMCIPHER, outPublic.t.publicArea.unique.sym.t))
-            return -5;
-    }
-    else
-    {
-        printf("\nThe algorithm type(0x%4.4x) is not supported\n", outPublic.t.publicArea.type);
-        return -6;
+        case TPM_ALG_RSA:
+            sizedBuffer = &outPublic.t.publicArea.unique.rsa.b;
+            /* no break */
+        case TPM_ALG_KEYEDHASH:
+            sizedBuffer = &outPublic.t.publicArea.unique.keyedHash.b;
+            /* no break */
+        case TPM_ALG_SYMCIPHER:
+            sizedBuffer = &outPublic.t.publicArea.unique.sym.b;
+            rc = SaveKeyToFile(akKeyFile, outPublic.t.publicArea.type, sizedBuffer) ? -2 : 0;
+        break;
```

Comment:
You should not do it this way, this will not function as your expectation: no matter outPublic.t.publicArea.type is TPM_ALG_RSA, TPM_ALG_KEYEDHASH, or TPM_ALG_SYMCIPHER, the final result will be sizedBuffer = &outPublic.t.publicArea.unique.sym.b;